File: backend/src/models/budget_member.rs
// Copyright (c) 2024 Tempo Budget
// SPDX-License-Identifier: MIT
//
// Budget member data models for group budget functionality.

//! # Budget Member Models
//!
//! This module defines data structures for managing membership in group budgets,
//! including member roles and invitation requests.

use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use utoipa::ToSchema;

/// Represents a member of a group budget.
#[derive(Debug, Clone, Serialize, Deserialize, FromRow, ToSchema)]
pub struct BudgetMember {
    /// Unique identifier (UUID)
    pub id: String,
    /// ID of the budget
    pub budget_id: String,
    /// ID of the user who is a member
    pub user_id: String,
    /// Role of the member: "owner", "admin", or "member"
    pub role: String,
    /// Timestamp when the membership was created (RFC 3339 format)
    pub created_at: String,
}

/// Request payload for inviting a member to a budget.
#[derive(Debug, Deserialize, ToSchema)]
pub struct InviteMemberRequest {
    /// Email address of the user to invite
    pub email: String,
    /// Role to assign: "admin" or "member" (owner cannot be assigned)
    pub role: String,
}

/// Budget member with associated user information.
///
/// This structure is used when displaying member lists with user details.
#[derive(Debug, Clone, Serialize, Deserialize, FromRow, ToSchema)]
pub struct BudgetMemberWithUser {
    /// Unique identifier (UUID)
    pub id: String,
    /// ID of the budget
    pub budget_id: String,
    /// ID of the user
    pub user_id: String,
    /// Role of the member
    pub role: String,
    /// Timestamp when the membership was created (RFC 3339 format)
    pub created_at: String,
    /// User's display name
    pub user_name: String,
    /// User's email address
    pub user_email: String,
    /// User's avatar URL (optional)
    pub user_avatar: Option<String>,
}

File: backend/app/models/budget_member.py
# Copyright (c) 2024 Tempo Budget
# SPDX-License-Identifier: MIT
"""Budget member data models."""

from typing import Literal
from pydantic import BaseModel, EmailStr, Field


class BudgetMember(BaseModel):
    """Represents a member of a group budget."""
    id: str = Field(..., description="Unique identifier (UUID)")
    budget_id: str = Field(..., description="ID of the budget")
    user_id: str = Field(..., description="ID of the user")
    role: str = Field(..., description="Role: 'owner' or 'member'")
    share: float = Field(50.0, description="Percentage share of the budget (0-100)")
    created_at: str = Field(..., description="Creation timestamp")

    class Config:
        from_attributes = True


class BudgetMemberWithUser(BaseModel):
    """Budget member with associated user information."""
    id: str = Field(..., description="Unique identifier (UUID)")
    budget_id: str = Field(..., description="ID of the budget")
    user_id: str = Field(..., description="ID of the user")
    role: str = Field(..., description="Role of the member")
    share: float = Field(50.0, description="Percentage share of the budget (0-100)")
    created_at: str = Field(..., description="Creation timestamp")
    user_name: str = Field(..., description="User's display name")
    user_email: str = Field(..., description="User's email address")
    user_avatar: str | None = Field(None, description="User's avatar URL")


class UpdateMemberShareRequest(BaseModel):
    """Request payload for updating a member's share."""
    share: float = Field(..., ge=0, le=100, description="New share percentage (0-100)")


class MemberBalance(BaseModel):
    """Balance information for a budget member."""
    user_id: str = Field(..., description="ID of the user")
    user_name: str = Field(..., description="User's display name")
    share: float = Field(..., description="Percentage share of the budget")
    total_due: float = Field(..., description="Total amount due based on share")
    total_paid: float = Field(..., description="Total amount paid by this member")
    balance: float = Field(..., description="Balance (positive = overpaid, negative = underpaid)")


class InviteMemberRequest(BaseModel):
    """Request payload for inviting a member to a budget."""
    email: EmailStr = Field(..., description="Email address of the user to invite")
    role: Literal["member"] = Field("member", description="Role to assign")
